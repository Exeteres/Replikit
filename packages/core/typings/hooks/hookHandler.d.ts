import { HookMap } from "@replikit/core/typings";

export type HookName = keyof HookMap;

export type HookHandler<T extends HookName = HookName> = (
    payload: HookMap[T]
) => Promise<void> | void;
