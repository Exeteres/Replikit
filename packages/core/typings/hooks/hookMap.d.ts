import { ResolvedConfiguration } from "@replikit/core/typings";

export interface HookMap {
    "core:startup:init": void;
    "core:startup:done": void;
    "core:startup:ready": void;
    "core:shutdown:init": void;
    "core:shutdown:done": void;
    "core:settings:update": ResolvedConfiguration;
}
