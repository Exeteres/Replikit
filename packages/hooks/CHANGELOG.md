# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.17.0](https://github.com/Exeteres/Replikit/compare/v0.16.0...v0.17.0) (2020-08-21)

**Note:** Version bump only for package @replikit/hooks





# [0.16.0](https://github.com/Exeteres/Replikit/compare/v0.15.1...v0.16.0) (2020-08-18)

**Note:** Version bump only for package @replikit/hooks





## [0.15.1](https://github.com/Exeteres/Replikit/compare/v0.15.0...v0.15.1) (2020-08-13)

**Note:** Version bump only for package @replikit/hooks





# [0.15.0](https://github.com/Exeteres/Replikit/compare/v0.14.0...v0.15.0) (2020-08-13)

**Note:** Version bump only for package @replikit/hooks





# [0.14.0](https://github.com/Exeteres/Replikit/compare/v0.13.0...v0.14.0) (2020-08-05)

**Note:** Version bump only for package @replikit/hooks





# [0.13.0](https://github.com/Exeteres/Replikit/compare/v0.12.1...v0.13.0) (2020-08-04)


### Features

* **hooks:** Add hooks to work with storage parameters ([09536f2](https://github.com/Exeteres/Replikit/commit/09536f2082364c46a8c14b70694e08e4b27e0b8e))





## [0.12.1](https://github.com/Exeteres/Replikit/compare/v0.12.0...v0.12.1) (2020-07-30)

**Note:** Version bump only for package @replikit/hooks





# [0.12.0](https://github.com/Exeteres/Replikit/compare/v0.11.2...v0.12.0) (2020-07-30)


### Features

* **hooks:** Add useChannel hook ([265c4de](https://github.com/Exeteres/Replikit/commit/265c4de8f17ea1db9c63e8900be28e1c50e5ae14))





## [0.11.2](https://github.com/Exeteres/Replikit/compare/v0.11.1...v0.11.2) (2020-07-24)

**Note:** Version bump only for package @replikit/hooks





## [0.11.1](https://github.com/Exeteres/Replikit/compare/v0.11.0...v0.11.1) (2020-07-24)

**Note:** Version bump only for package @replikit/hooks






# [0.11.0](https://github.com/Exeteres/Replikit/compare/v0.10.1...v0.11.0) (2020-07-24)


### Features

* **hooks:** Add basic hook implementation ([d786346](https://github.com/Exeteres/Replikit/commit/d786346d18eba2476794dbc165f92d9b405500a8))
* **hooks:** Add useLocale hook ([927f716](https://github.com/Exeteres/Replikit/commit/927f7165d4217aaad27eea2275f8668b1a8dc906))
* **hooks:** Redesign hooks system ([362a403](https://github.com/Exeteres/Replikit/commit/362a403fac7311a79a43f1d75346edcdfbb20a11))
