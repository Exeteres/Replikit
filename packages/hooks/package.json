{
    "name": "@replikit/hooks",
    "version": "0.11.1",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "plugin",
        "typings"
    ],
    "scripts": {
        "build:package": "tsc -p tsconfig.build.json",
        "build:plugin": "tsc -p tsconfig.plugin.json",
        "build": "yarn build:plugin && yarn build:package"
    },
    "publishConfig": {
        "access": "public"
    },
    "dependencies": {
        "@replikit/commands": "^0.11.1",
        "@replikit/core": "^0.11.1"
    }
}
