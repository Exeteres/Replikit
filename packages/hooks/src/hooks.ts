/* eslint-disable @typescript-eslint/no-unused-vars */

import { Constructor } from "@replikit/core/typings";
import { ParameterOptions, CommandContext, RestParameterOptions } from "@replikit/commands/typings";
import { useContext } from "@replikit/hooks";
import { NormalizeType, TextParameterOptions } from "@replikit/commands";
import { LocaleConstructor } from "@replikit/i18n/typings";

export function useCommandContext(): CommandContext {
    return useContext() as CommandContext;
}

export function useRequired<T>(
    name: string,
    type: Constructor<T>,
    options?: ParameterOptions<T>
): NormalizeType<T> {
    const context = useCommandContext();
    return context.params[name] as NormalizeType<T>;
}

export function useOptional<T>(
    name: string,
    type: Constructor<T>,
    options: ParameterOptions<T> & { default: T }
): NormalizeType<T>;

export function useOptional<T>(
    name: string,
    type: Constructor<T>,
    options?: ParameterOptions<T>
): NormalizeType<T> | undefined;

export function useOptional<T>(
    name: string,
    type: Constructor<T>,
    options?: ParameterOptions<T>
): NormalizeType<T> | undefined {
    const context = useCommandContext();
    return context.params[name] as NormalizeType<T>;
}

export function useText(name?: string, options?: TextParameterOptions): string;
export function useText(options: TextParameterOptions): string;

export function useText(
    name?: string | TextParameterOptions,
    options?: TextParameterOptions
): string {
    const context = useCommandContext();
    const propName = typeof name === "string" ? name : "text";
    return context.params[propName] as string;
}

export function useRest<T>(
    name: string,
    type: Constructor<T>,
    options: RestParameterOptions<T>
): NormalizeType<T>[] {
    const context = useCommandContext();
    return context.params[name] as NormalizeType<T>[];
}

export function useLocale<T>(type: LocaleConstructor<T>): T {
    const context = useContext();
    return context.getLocale(type);
}
