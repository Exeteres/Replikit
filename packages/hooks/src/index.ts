export * from "./utils";
export * from "./contextStorage";
export * from "./hooks";
