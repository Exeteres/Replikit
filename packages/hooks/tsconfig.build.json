{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/hooks": ["hooks/src"]
        }
    },
    "include": ["src"],
    "exclude": ["src/plugin"]
}
