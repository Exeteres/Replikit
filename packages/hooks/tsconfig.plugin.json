{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "plugin",
        "rootDir": "src/plugin",
        "paths": {
            "@replikit/hooks": ["hooks/src"]
        },
        "plugins": []
    },
    "include": ["src/plugin"]
}
