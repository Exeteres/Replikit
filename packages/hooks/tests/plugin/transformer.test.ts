import hookTransformer, { FunctionNotModifiedError } from "@replikit/hooks/plugin";
import { code, createTranspiler } from "@replikit/test-utils";

const transpileModule = createTranspiler({ before: [hookTransformer] });

describe("hook transformer", () => {
    it("should collect hook calls from function body and add corresponding applyHook calls to CommandBuilder", () => {
        const input = code`
            command("test")
                .handler(handler)
                .register();

            function handler() {
                const p1 = useRequired("p1", String);
                const p2 = useOptional("p2", Number, { default: 123 });
            }
        `;
        const result = transpileModule(input);
        expect(result).toMatchSnapshot();
    });

    it("should do not require @replikit/hooks if it already required", () => {
        const input = code`
            import { useRequired } from "@replikit/hooks";

            command("test")
                .handler(handler)
                .register();

            function handler() {
                const p1 = useRequired("p1", String);
            }
        `;
        const result = transpileModule(input);
        expect(result).toMatchSnapshot();
    });

    it("should throw an error when declaring a handler before CommandBuilder call", () => {
        const input = code`
            function handler() {
                const p1 = useRequired("p1", String);
            }

            command("test")
                .handler(handler)
                .register();
        `;
        const action = () => transpileModule(input);
        expect(action).toThrow(FunctionNotModifiedError);
    });

    it("should ignore arrow function as handlers", () => {
        const input = code`
            command("test")
                .handler(() => {})
                .register();
        `;
        const result = transpileModule(input);
        expect(result).toMatchSnapshot();
    });

    it("should ignore commands without handler", () => {
        const input = code`
            command("test")
                .commands()
                .register();
        `;
        const result = transpileModule(input);
        expect(result).toMatchSnapshot();
    });
});
