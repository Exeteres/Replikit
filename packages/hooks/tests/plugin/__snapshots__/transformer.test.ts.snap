// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hook transformer should collect hook calls from function body and add corresponding applyHook calls to CommandBuilder 1`] = `
"var hooks_1 = require(\\"@replikit/hooks\\");
hooks_1.applyHook(hooks_1.applyHook(command(\\"test\\"), useRequired, [\\"p1\\", String]), useOptional, [\\"p2\\", Number, { \\"default\\": 123 }]).handler(handler).register();
function handler() {
    hooks_1.setContext(arguments[0]);
    var p1 = useRequired(\\"p1\\", String);
    var p2 = useOptional(\\"p2\\", Number, { \\"default\\": 123 });
}
"
`;

exports[`hook transformer should do not require @replikit/hooks if it already required 1`] = `
"\\"use strict\\";
exports.__esModule = true;
var hooks_1 = require(\\"@replikit/hooks\\");
hooks_1.applyHook(command(\\"test\\"), hooks_1.useRequired, [\\"p1\\", String]).handler(handler).register();
function handler() {
    hooks_1.setContext(arguments[0]);
    var p1 = hooks_1.useRequired(\\"p1\\", String);
}
"
`;

exports[`hook transformer should ignore arrow function as handlers 1`] = `
"command(\\"test\\")
    .handler(function () { })
    .register();
"
`;

exports[`hook transformer should ignore commands without handler 1`] = `
"command(\\"test\\")
    .commands()
    .register();
"
`;
